### PR TITLE
Balance streaming parents

### PR DESCRIPTION
**Pull request by ktsaou**
## Generic code base improvements
- [x] abstracted `struct destinations` to `STREAM_PARENT` and created an API without exposing internal structures
- [x] split stream_conf into its source files
- [x] split libnetdata/sockets into multiple source files
- [x] organized `rrdhost` in `rrd.h` about streaming
- [x] introduction of `ND_SOCK` a generic object for handling all kinds of sockets, transparently applying TLS encryption when required - this simplified the rest of the code tremendously, since now TLS is part of the transport transparently.
- [x] streaming now uses `ND_SOCK` everywhere (receivers and senders)
- [x] global statistics is now named `telemetry` and is split into multiple files
- [x] netdata.conf parsing in `main.c` is now split into multiple files matching the section names of the config file

## Stream Info and Balancing
- [x] added `/api/v3/stream_info?machine_guid=UUID` which exposes information about the current load on a parent (nodes, receivers), retention about the child (first_time, last_time), and status information.
- [x] added balancing when connecting to parents (uses the `stream_info` api of the parents when available).
- [ ] rebalance parents after replication finishes on a parent

### Streaming
- [x] randomized delays and postpone times to prevent huge stress on busy parents on startup or reconnects
- [x] permanently ban parents that are the origin of a host (this prevents clusters from trying all the time each other for the origin host)
- [x] temporary ban parents that are before us in the stream path (we receive data from/via them) - there is no point to retry to connect to them, until the receiver exits.
- [x] temporary ban parents that timeout, to allow all the nodes to connect quickly.

### Reducing CPU load
- [x] streaming now has a fixed number of threads, up to the number of CPU cores.
- [x] hosts are distributed to the streaming threads evenly, so if the number of hosts is more than the number of cores, each thread will serve more than 1 host.
- [x] receivers and senders are served by the same thread for the same host. This reduces significantly cross CPU core atomic collisions and speeds up performance.
- [x] a separate connector thread is responsible for connecting senders to their parents. A single thread for all nodes. This thread bans parents that timeout to avoid delays while connecting.

###
- [x] out of memory protection (leave by default 10% of the total RAM free)
- [x] use all ram for cache option

### ML related
- [x] disable prediction and training when anomaly information is received via streaming (added new capability `ML_MODELS`)
- [x] propagate trained ML models with streaming (waiting for @vkalintiris to push here)

---

## Fixed Bugs (v2 stable, minor release required)

- [x] fixed a bug in replication during shutdown, referencing a NULL pointer
- [x] fixed a bug in streaming, when the receiver exits the sender buffer may still be in-use due to an incomplete begin-set-end v2
- [x] fixed a bug in rrdhost_status, where a `!rrdhost_flag_check(host, RRDHOST_FLAG_RRDPUSH_RECEIVER_DISCONNECTED)` was incorrectly `rrdhost_flag_set(host, RRDHOST_FLAG_RRDPUSH_RECEIVER_DISCONNECTED)`

### CODE REVIEW

File: src/streaming/stream-conf.c [hunk 1]
```diff
@@ -1,28 +1,46 @@
...
-// to have the remote netdata re-sync the charts
-// to its current clock, we send for this many
-// iterations a BEGIN line without microseconds
-// this is for the first iterations of each chart
-unsigned int stream_conf_initial_clock_resync_iterations = 60;
+#include "daemon/common.h"
+
+static struct config stream_config = APPCONFIG_INITIALIZER;
+
+struct _stream_send stream_send = {
```
Comment (vkalintiris) [inline src/streaming/stream-conf.c:7]: @ktsaou types/variables starting with a leading underscored are reserved for the runtime, do not use such names.

Comment (ktsaou) [reply src/streaming/stream-conf.c:7]: You have to open windows.h

You will be impressed! Almost every struct and enum starts with underscore... 

I will rename them.

**Comment by ktsaou** [comment]
> This does not seem to work when I setup a child/parent from this branch. I can see that the models are properly serialized on the child's side, but they are not received on the parent at all. `netdata.ml_models_streamed` in the "Netdata -> machine learning" section is the relevant chart we can use to track this.

@vkalintiris the json keyword was wrong. It had STREAM_PATH, not ML_MODEL.

**Comment by vkalintiris** [comment]
> > This does not seem to work when I setup a child/parent from this branch. I can see that the models are properly serialized on the child's side, but they are not received on the parent at all. `netdata.ml_models_streamed` in the "Netdata -> machine learning" section is the relevant chart we can use to track this.
> 
> @vkalintiris the json keyword was wrong. It had STREAM_PATH, not ML_MODEL.

Thanks, I'll do more testing then.

**Comment by vkalintiris** [comment]
@ktsaou 

There are three optimizations which can be reverted independently, through a single commit:

- [Batch gorilla writes only from writers](https://github.com/netdata/netdata/pull/18945/commits/e8a54ab895fcc84321515c1b93f7542ae1ce7e45)
- [Use re2c for the line splitting pluginsd](https://github.com/netdata/netdata/pull/18945/commits/2b2f9d38877780f3058727c343e67b902c5b615d)
- [make buffered_reader_next_line() a couple times faster
](https://github.com/netdata/netdata/pull/18945/commits/d692ea86b201fd3e4ff5445601c7372e69240cfb)

In my setup - 200 children pushing to a parent - everything works correctly. However, I noticed that periodically some children disconnect and reconnect after a while.

I'm getting the same behaviour even after reverting my commits.

Screenshot 2024-11-27 at 12 36 15

**Pull request merged**